Tree Data Structures/Trie Tree/Java/Trie.java

```diff
@@ -0,0 +1,216 @@
+/* ===== ===== =====
+
+Theory of Programming
+Trie Tree Implementation
+http://theoryofprogramming.com/2015/01/16/trie-tree-implementation/
+GitHub - https://github.com/VamsiSangam/theoryofprogramming
+Code Contributor - Saurabh Rane (github.com/ranesr)
+
+===== ===== ===== */
+
+import java.util.HashMap;
+
+public class Trie {
+
+	// root node
+	private TrieNode root;
+	
+	// default case
+	public Trie() {
+		root = new TrieNode();
+		root.setParent(new TrieNode());
+	}
+	
+	// inserting a word	
+	public void insert(String word) {
+		HashMap<Character, TrieNode> children = root.getChildren();
+		
+		// parent node
+		TrieNode p = root;
+		
+		for(int i = 0; i < word.length(); i++) {			
+			char c = word.charAt(i);
+			
+			TrieNode t = null;
+			
+			if(children.containsKey(c)) {
+				
+				// character already found in children
+				t = children.get(c);
+				
+			} else {
+				
+				// adding new character to children				
+				t = new TrieNode(c);
+				children.put(c, t);
+				t.setParent(p);
+			}
+			
+			p = t;
+			children = t.getChildren();
+			
+			// last character from word
+			if(i == word.length()-1) {
+				t.setLeaf(true);
+			}
+		}
+	}
+	
+	public boolean delete(String word) {
+		TrieNode t = searchNode(word);
+		
+		// word is present if t is not null & is leaf as t is a trie node for last character from word
+		if(t != null && t.isLeaf()) {
+			
+			TrieNode p = t.getParent();
+			char c = t.getCharacter();
+
+			// removing children of parent if only one child is present
+			while(p.getChildren().size() == 1) {
+				p.resetChildren();
+				t = p;
+				c = t.getCharacter();
+				p = t.getParent();
+			}
+			
+			// removing the remaining character
+			char delete = c;
+			p.getChildren().entrySet().removeIf(e -> e.getKey().equals(delete));
+
+			return true;
+		}
+		return false;
+	}
+	
+	public boolean search(String word) {
+		TrieNode t = searchNode(word);
+		
+		// word is present if t is not null & is leaf as t is a trie node for last character from word
+		if(t != null && t.isLeaf()) {
+			return true;
+		}
+		
+		return false;
+	}
+	
+	public boolean startsWith(String prefix) {
+		HashMap<Character, TrieNode> children = root.getChildren();
+		
+		TrieNode t = null;
+		
+		for(int i = 0; i < prefix.length(); i++) {
+			char c = prefix.charAt(i);
+			if(children.containsKey(c)) {
+			
+				t = children.get(c);
+				children = t.getChildren();
+
+			} else {
+				// character mismatch
+				return false;
+			}
+		}
+
+		// reached the end of the prefix word
+		return true;
+	}
+	
+	// helper method for searching & startsWith functions
+	public TrieNode searchNode(String word) {
```

Comment:
Please make the helper method ```private``` and the comment should be saying ```helper method for searching & delete functions``` as the ```startsWith``` function doesn't seem to be using this method.  Please modify the comment.